dffml/df/base.py

```diff
@@ -215,6 +216,15 @@ def wrap(func):
                             param.annotation, param.annotation.__name__
                         ),
                     )
+                elif is_dataclass(param.annotation) or bool(
+                    issubclass(param.annotation, tuple)
+                    and hasattr(param.annotation, "_asdict")
+                ):
+                    kwargs["inputs"][name] = Definition(
+                        name=".".join(name_list),
+                        primitive="map",
+                        spec=param.annotation,
+                    )
```

Comment:
When you implment the same thing but subspec, you'll have an if statement after the one you added where you check if the `__origin__` (use typing.get_origin, also in base.py but you would probably move it to util.data). If origin is Typing.list then set primitive to array, otherwise set it to map if dict. If it's neither of those, then fall through to the last case.

You'll still want to have the logic you added in this if statement to validate that the list / dict has a inner type which is a Named Tuple or dataclass

Reply:
You'll also want to modify the code in Input.__init__ so that it checks what it's primitive is, and throws an error if the datatype passed isn't right type, so if it's not a list or dict, rather than the current behavior which just checks if the value is the correct.

Reply:
thanks, done